keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,842 @@
+  containers running in the host. There is not support for [multiple ID
+  mappings](https://github.com/moby/moby/issues/28593) yet. Updating dockershim
+  is out of scope of this KEP as [it's going to be
+  deprecated]((https://github.com/kubernetes/enhancements/pull/1985/)) and it
+  offers a very limited support for user namespaces.
+- **containerd**:
+  It's quite straightforward to implement the CRI changes proposed below in
+  containerd/cri, we did it in
+  [this](https://github.com/kinvolk/containerd-cri/commits/mauricio/userns_poc)
+  PoC.
+- **cri-o**:
+  CRI-O recently [added](https://github.com/cri-o/cri-o/pull/3944) support for
+  user namespaces through a pod annotation. The extensions to make it work with
+  the CRI changes proposed here are small.
+- gVisor, katacontainers: Yet to be investigated.
+
+containerd and cri-o will provide support for the 3 possible values of `userNamespaceMode`.
+
+### Risks and Mitigations
+
+#### Breaking Existing Workloads
+
+Some features that don't work when the host user namespace is not shared are:
+
+- **Some Capabilities**:
+  The Linux kernel takes into consideration the user namespace a process is
+  running in while performing the capabilities check. There are some
+  capabilities that are only available in the initial (host) user namespace such
+  as `CAP_SYS_TIME`, `CAP_SYS_MODULE` & `CAP_MKNOD`.
+
+  If a pod is given one of those capabilities it will still be deployed, but the
+  capability will be ineffective and processes using those capabilities will
+  fail. This is not impacting the implementation in Kubernetes. If users need
+  the capability to be effective, they should use `userNamespaceMode=Host`.
+
+  The list of such capabilities is likely to change from one Linux version to
+  another. For example, Linux now has [time
+  namespaces](https://man7.org/linux/man-pages/man7/time_namespaces.7.html) and
+  there are ways to make `CAP_SYS_TIME` work inside a user namespace. There are
+  also discussions to make `CAP_MKNOD` work in user namespaces.
+
+- **Sharing Host Namespaces**:
+  There are some limitations in the Linux kernel and in the runtimes that
+  prevent sharing other host namespaces when the host user namespace is not
+  shared.
+  - Mounting `mqueue` (`/dev/mqueue`) is not allowed from a process in a user
+    namespace that does not own the IPC namespace. Pods with `hostIPC=true` and
+    `userNamespaceMode=Pod|Cluster` can fail.
+  - Mounting `procfs` (`/proc`) is not allowed from a process in a user
+    namespace that does not own the PID namespace. Pods with `hostPID=true` and
+    `userNamespaceMode=Pod|Cluster` can fail.
+  - Mounting `sysfs` (`/sys`) is not allowed from a process in a user namespace
+    that does not own the network namespace. Impact: pods with
+    `hostNetwork=true` and `userNamespaceMode=Pod|Cluster` can fail.
+
+  If users specify `userNamespaceMode=Pod|Cluster` and one of these
```

Comment:
Should we statically enforce this at the API layer?  Do we expect it to change over time?  Or should we just document all of these as "consult your OS docs to learn more about how these interact" ?